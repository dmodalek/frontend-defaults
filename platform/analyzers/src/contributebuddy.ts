import { directoryExists, getDependencyInstallation, DependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ContributebuddyAnalyzerResult = {
    contributebuddy: boolean;
    contributebuddyConfigFolder?: string,
    contributebuddyInstallation?: DependencyInstallation
};

export const contributebuddyAnalyzer = async (cwd: string): Promise<ContributebuddyAnalyzerResult> => {
    const CONTRIBUTE_BUDDY_FOLDER = join(cwd, '.contributebuddy');
    const contributebuddyConfigFolder = await directoryExists(CONTRIBUTE_BUDDY_FOLDER);
    const contributebuddyInstallation = await getDependencyInstallation(cwd, 'contribute-buddy');

    if (contributebuddyConfigFolder || contributebuddyInstallation.declared) {
        return {
            contributebuddy: true,
            contributebuddyConfigFolder: CONTRIBUTE_BUDDY_FOLDER,
            contributebuddyInstallation

        }
    }

    return {
        contributebuddy: false,
    };
}