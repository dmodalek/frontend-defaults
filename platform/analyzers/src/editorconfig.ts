import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type EditorconfigAnalyzerResult = {
	editorConfig: boolean;
};

export const editorConfigAnalyzer = async (cwd: string): Promise<EditorconfigAnalyzerResult> => {
	return {
		editorConfig: await fileExists(join(cwd, '.editorconfig'))
	}
}