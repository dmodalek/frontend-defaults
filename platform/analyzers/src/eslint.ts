import { DependencyInstallation, fileExists, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ESLintAnalyzerResult = {
	eslint: boolean;
	eslintIgnore?: boolean;
	eslintInstallation?: DependencyInstallation
};

export const eslintAnalyzer = async (cwd: string): Promise<ESLintAnalyzerResult> => {
	const doesESLintExist = await fileExists(join(cwd, '.eslintrc.js'));

	if (doesESLintExist) {
		const doesESLintIgnoreExist = await fileExists(join(cwd, '.eslintignore'));
		const eslintInstallation = await getDependencyInstallation(cwd, 'eslint');

		return {
			eslint: doesESLintExist,
			eslintIgnore: doesESLintIgnoreExist,
			eslintInstallation
		};
	}

	return {
		eslint: doesESLintExist,
	};
}