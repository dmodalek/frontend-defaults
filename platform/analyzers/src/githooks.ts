import {
	DependencyInstallation,
	fileExists,
	getDependencyInstallation,
	getPackageJSON,
	IPackage,
} from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitHooksLocation = 'rc' | 'package';

export type GitHooksAnalyzerResult = {
	gitHooks: boolean;
	gitHooksInstallation?: DependencyInstallation;
	gitHooksLocation?: GitHooksLocation;
};

type PackageWithHuskyConfig = IPackage & {
	husky: any;
};

export const gitHooksAnalyzer = async (cwd: string): Promise<GitHooksAnalyzerResult> => {
	const doesGitHooksConfigExist = await fileExists(join(cwd, '.huskyrc'));
	const gitHooksInstallation = await getDependencyInstallation(cwd, 'husky');
	const pkg: PackageWithHuskyConfig = (await getPackageJSON(cwd)) as any;

	if (doesGitHooksConfigExist || pkg.husky) {
		return {
			gitHooks: doesGitHooksConfigExist,
			gitHooksLocation: pkg.husky ? 'package' : 'rc',
			gitHooksInstallation,
		};
	}

	return {
		gitHooks: false,
	};
};
