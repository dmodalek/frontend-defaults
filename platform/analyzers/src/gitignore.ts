import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitIgnoreAnalyzerResult = {
	gitignore: boolean;
};

export const gitIgnoreAnalyzer = async (cwd: string): Promise<GitIgnoreAnalyzerResult> => {
	const doesGitignoreExist = await fileExists(join(cwd, '.gitignore'));

	return {
		gitignore: doesGitignoreExist,
	};
}