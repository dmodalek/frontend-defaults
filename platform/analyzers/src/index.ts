// export all related analyzer stuff for contributebuddy
export { contributebuddyAnalyzer, ContributebuddyAnalyzerResult } from './contributebuddy';

// export all related analyzer stuff for editorconfig
export { editorConfigAnalyzer, EditorconfigAnalyzerResult } from './editorconfig';

// export all related analyzer stuff for eslint
export { eslintAnalyzer, ESLintAnalyzerResult } from './eslint';

// export all related analyzer stuff for githooks
export { gitHooksAnalyzer, GitHooksAnalyzerResult } from './githooks';

// export all related analyzer stuff for gitignore
export { gitIgnoreAnalyzer, GitIgnoreAnalyzerResult } from './gitignore';

// export all related analyzer stuff for license
export { licenseAnalyzer, LicenseAnalyzerResult } from './license';

// export all related analyzer stuff for nitro
export { nitroAnalyzer, NitroAnalyzerResult } from './nitro';

// export all related analyzer stuff for node
export { nodeAnalyzer, NodeAnalyzerResult } from './node';

// export all related analyzer stuff for npmrc
export { npmrcAnalyzer, NPMRCAnalyzerResult } from './npmrc';

// export all related analyzer stuff for prettier
export { prettierAnalyzer, PrettierAnalyzerResult } from './prettier';

// export all related analyzer stuff for readme
export { readmeAnalyzer, ReadmeAnalyzerResult } from './readme';

// export all related analyzer stuff for repository
export { repositoryAnalyzer, RepositoryAnalyzerResult } from './repository';

// export all related analyzer stuff for stylelint
export { stylelintAnalyzer, StylelintAnalyzerResult } from './stylelint';

// export all related analyzer stuff for tslint
export { tslintAnalyzer, TSLintAnalyzerResult } from './tslint';

// export all related analyzer stuff for typescript
export { typescriptAnalyzer, TypeScriptAnalyzerResult } from './typescript';

// export all related analyzer stuff for webpack
export { webpackAnalyzer, WebpackAnalyzerResult } from './webpack';
