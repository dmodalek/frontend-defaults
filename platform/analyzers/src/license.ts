import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type LicenseAnalyzerResult = {
	license: boolean;
};

export const licenseAnalyzer = async (cwd: string): Promise<LicenseAnalyzerResult> => {
	// TODO: Recognize also license or License (case) + type from package
	const licenseExists = await fileExists(join(cwd, 'LICENSE'));

	return {
		license: licenseExists,
	};
}