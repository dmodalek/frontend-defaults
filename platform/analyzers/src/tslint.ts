import { join } from 'path';
import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';

export type TSLintAnalyzerResult = {
	tslint: boolean;
	tslintInstallation?: false | string;
};

export class TSLintAnalyzer extends Analyzer<TSLintAnalyzerResult> {
	async analyze(): Promise<TSLintAnalyzerResult> {
		const doesTSLintExist = await fileExists(join(this.context, 'tslint.json'));

		if (doesTSLintExist) {
			const eslintInstallation = this.packageAnalyzer.anyDependencyExists('tslint');

			return {
				tslint: doesTSLintExist,
				tslintInstallation: eslintInstallation ? eslintInstallation.version : false,
			};
		}

		return {
			tslint: doesTSLintExist,
		};
	}
}
