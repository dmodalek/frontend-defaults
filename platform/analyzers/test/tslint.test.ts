import { getFixtureContext } from './utils';
import { tslintAnalyzer } from '../src/tslint';

const FIXTURE = getFixtureContext('tslint-project');
const FIXTURE_NOT_INSTALLED = getFixtureContext('tslint-project-not-installed');
const FIXTURE_NOT_USED = getFixtureContext('default-project');

describe('Analyzers', () => {
	describe('TSLintAnalyzer', () => {
		it('should not crash', async () => {
			expect(async () => await tslintAnalyzer(FIXTURE)).not.toThrow();
		});

		it('should analyze a project without tslint correctly', async () => {
			const analytics = await tslintAnalyzer(FIXTURE_NOT_USED);

			expect(analytics.tslint).toEqual(false);
			expect(analytics).toMatchSnapshot();
		});

		it('should analyze a project with installation correctly', async () => {
			const analytics = await tslintAnalyzer(FIXTURE);

			expect(analytics.tslint).toEqual(true);
			expect(analytics.tslintInstallation!.declared).toEqual('5.12.1');
			expect(analytics.tslintInstallation!.installed).toBeFalsy(); // FIXME: not mocked yet
			expect(analytics).toMatchSnapshot({
				tslintInstallation: {
					"latest": expect.any(String)
				}
			});
		});

		it('should analyze a project without installation correctly', async () => {
			const analytics = await tslintAnalyzer(FIXTURE_NOT_INSTALLED);

			expect(analytics.tslint).toEqual(true);
			expect(analytics.tslintInstallation!.installed).toBeFalsy();
			expect(analytics).toMatchSnapshot({
				tslintInstallation: {
					"latest": expect.any(String)
				}
			});
		});
	});
});
