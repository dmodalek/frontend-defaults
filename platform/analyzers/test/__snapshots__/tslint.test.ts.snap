// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzers TSLintAnalyzer should analyze a project with installation correctly 1`] = `
Object {
  "tslint": true,
  "tslintInstallation": Object {
    "declared": "5.12.1",
    "installed": null,
    "latest": "5.12.1",
    "location": "devDependencies",
  },
}
`;

exports[`Analyzers TSLintAnalyzer should analyze a project without installation correctly 1`] = `
Object {
  "tslint": true,
  "tslintInstallation": Object {
    "declared": null,
    "installed": null,
    "latest": "5.12.1",
    "location": null,
  },
}
`;

exports[`Analyzers TSLintAnalyzer should analyze a project without tslint correctly 1`] = `
Object {
  "tslint": false,
}
`;
